### DISCUSSION TITLE
Enhance the moleculer ecosystem

### DISCUSSION THREAD

**Discussion by ishfx** [moleculerjs/moleculer#1204] [Show and tell]
## Goal
The purpose of this discussion is to address the potential obsolescence concern associated with Moleculer and explore ways to update the framework with modern JavaScript features. If needed, we can move this discussion to a more appropriate platform.

## Is your feature request related to a problem? Please describe.
MoleculerJS is undoubtedly one of the most powerful microservices frameworks for Node.js, but one of the reasons it may not be gaining as much traction as it could is due to its reliance on old JavaScript syntax. This can be a turn-off for developers who prefer modern TypeScript features and syntax. Furthermore, the modern JavaScript ecosystem has evolved significantly in recent years, and in order to stay relevant and appeal to a wider audience, MoleculerJS needs to embrace this direction. 

## Describe the solution you'd like
I want to introduce a new library that I believe will greatly enhance the development experience with MoleculerJS. My library, called  `moleculer-plus`, provides a set of TypeScript decorators that seamlessly integrate with MoleculerJS to allow developers to easily add modern TypeScript features and syntax to their microservices architecture without changing MoleculerJS's core code.

With `moleculer-plus`, developers can write cleaner and more concise code with the assurance that it will be validated using the fastest-validator library. The decorators also make it easy to declare services, actions, events, and APIs, with built-in support for validation rules.

I believe `moleculer-plus` will bring significant value to the moleculer ecosystem and help them streamline their development workflows. I would love the opportunity to collaborate and explore how we can integrate it into MoleculerJS.

Project link: https://github.com/ishfx/moleculer-plus

(fyi, this solution is being used in production)

**Comment by ishfx** [comment]
Hello everyone,

I was hoping to get some feedback and input from the community. However, so far, there hasn't been any activity on this discussion thread, and I would really appreciate it if more people could take a look and share their thoughts.

Moleculer is a fantastic project, and I believe that it has the potential to grow and improve even further with the help of the community's input and contributions. Whether you're a current user of moleculer or simply interested in the project, your feedback would be incredibly valuable to me and to the project as a whole.

So please, if you have a moment, take a look at the post and share your thoughts.

Thank you all for your time!

**Comment by zolero** [comment]
Hi @ishfx,

First off, thank you for introducing the moleculer-plus library. Your efforts towards enhancing the development experience for moleculer are genuinely appreciated.

While I acknowledge the benefits of TypeScript decorators for writing cleaner and more concise code, I have a couple of considerations:

1. **Backward Compatibility**: Ensuring minimal breaking changes for existing moleculer users is critical. We need to make sure they can upgrade to the moleculer ts version seamlessly.
3. **Decorator Setup**: Although powerful, decorators might complicate things for some users. Moleculers simplicity is its strength, so keeping decorators as an optional feature could be a beneficial approach.

We would like to be evolved with the transformations to a fully TypeScript moleculer as I think it could be powerful tool for not only the community how it is right now but for many more people. Let's get in contact to continue on these thoughts and maybe start a plan towards reaching this goal.

I'm available on Discord. @Username: **zolero**

Other TypeScript Discussion
_Feature Request - Complete TypeScript Implementation - https://github.com/moleculerjs/moleculer/discussions/1216_
